enhancements/machine-api/aws-imds-v2-support.md

```diff
@@ -0,0 +1,270 @@
+At this point, IMDSv2 enablement might be done manually via AWS console, but providing such settings on Machine API is a quite loud request from our end-users.
+
+### Goals
+
+- provide mechanisms to enable/enforce IMDSv2 for newly created machines via Machine API
+- provide mechanisms to enable/enforce IMDSv2 for master machines during installation procedure for newly created clusters
+
+### Non-Goals
+
+- perform automated migration between using IMDSv1 and IMDSv2 for existing infrastructure
+
+## Proposal
+
+- Add new options into install-config for the OCP installation program which will be passed as metadata service settings during virtual machines creation
+- Extend [AWSMachineProviderConfig](https://github.com/openshift/api/blob/1a6fa2913810101176a1d776f899fc4781b3fa50/machine/v1beta1/types_awsprovider.go#L12) by adding metadata service parameters
+- Pass metadata service parameters to AWS during new machines creation
+
+### User Stories
+
+- As an OpenShift cluster administrator, I would like to enforce IMDSv2 usage during an OCP cluster deployment procedure.
+- As an OpenShift cluster administrator, I would like to enforce IMDSv2 usage for newly created machines in my cluster via MachineSets configuration.
+- As an OpenShift cluster administrator, I would like to change IMDS interaction settings for newly created machines in my cluster.
+
+### API Extensions
+
+#### AWSMachineProviderConfig changes
+
+MachineAPI design mostly replicates parameters which [AWS API](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_InstanceMetadataOptionsRequest.html) consumes.
+
+```go
+type AWSMachineProviderConfig struct {
+    // Existing fields will not be modified
+    ...
+    // MetadataServiceOptions allows users to configure instance metadata service interaction options.
+    // If nothing specified, default AWS IMDS settings will be applied.
+    // https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_InstanceMetadataOptionsRequest.html
+    // +optional
+    MetadataServiceOptions *MetadataServiceOptions `json:"metadataServiceOptions,omitempty"`
+}
+```
+
+```go
+
+type InstanceMetadataHTTPTokens string
+
+const (
+    // InstanceMetadataHTTPTokensRequired enforces sending of a signed token header with any instance metadata retrieval (GET) requests.
+    // Enforces IMDSv2 usage.
+    InstanceMetadataHTTPTokensRequired = "required"
+    // InstanceMetadataHTTPTokensOptional allows IMDSv1 usage along with IMDSv2
+    InstanceMetadataHTTPTokensOptional = "optional"
+)
+
+// MetadataServiceOptions defines the options available to a user when configuring
+// Instance Metadata Service (IMDS) Options.
+type MetadataServiceOptions struct {
+    // HTTPTokens determines whether or not the host requires the use of authentication when interacting with the metadata service.
+    // When using authentication, this enforces v2 interaction method (IMDSv2) with the metadata service.
+    // Default: optional
```

Comment:
Is this going to be defaulted via a webhook or will this be a runtime default, do we see the default as likely to change over time?

Reply:
I planned to just omitted this and rely on AWS API defaults.

Reply:
In which case, we should update the wording here to match others (based on API conventions)
```suggestion
    // When omitted, this means the user has no opinion and the value is left to the platform to choose a good 
    // default, which is subject to change over time. The current default is optional.
```